File: video-controller/app/commandLine.js
import { exec } from 'child_process';
import { promises as fs } from 'fs';
import path from 'path';

/**
 * Utility class for command-line and file operations.
 */
export class CommandLine {
  constructor() {
    throw new Error('Utility class');
  }

  /**
   * Runs a shell command and returns stdout and stderr as a single string.
   * @param {string} command - The shell command to execute.
   * @returns {Promise<string>}
   */
  static runCommand(command) {
    return new Promise((resolve, reject) => {
      exec(command, (error, stdout, stderr) => {
        if (error) {
          reject(new Error(`Can't execute command: ${command}\n${stderr.trim()}`));
        } else {
          resolve((stdout + stderr).trim());
        }
      });
    });
  }

  /**
   * Deletes a file if it exists.
   * @param {string} fileName - Path to the file.
   * @param {boolean} throwIfNotExists - Whether to throw if the file doesn't exist.
   */
  static async fileDelete(fileName, throwIfNotExists = false) {
    try {
      await fs.unlink(fileName);
    } catch (err) {
      if (err.code === 'ENOENT' && !throwIfNotExists) {
        return;
      }
      throw new Error(`Can't delete file: ${fileName}\n${err.message}`);
    }
  }

  /**
   * Copies a file from one location to another.
   * @param {string} fileFrom - Source file path.
   * @param {string} fileTo - Destination file path.
   */
  static async fileCopy(fileFrom, fileTo) {
    try {
      await fs.copyFile(fileFrom, fileTo);
    } catch (err) {
      throw new Error(`Can't copy ${fileFrom} to ${fileTo}\n${err.message}`);
    }
  }

  /**
   * Checks if a file exists.
   * @param {string} fileName - Path to the file.
   * @returns {Promise<boolean>}
   */
  static async fileExists(fileName) {
    try {
      await fs.access(fileName);
      return true;
    } catch {
      return false;
    }
  }

  /**
   * Checks if a path is absolute.
   * @param {string} fileName - File path.
   * @returns {boolean}
   */
  static isAbsolute(fileName) {
    return path.isAbsolute(fileName);
  }
}

File: java/src/main/java/giis/selema/portable/CommandLine.java
package giis.selema.portable;

import java.io.BufferedReader;
import java.io.File;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.StringWriter;

import org.apache.commons.io.FileUtils;

public class CommandLine {

	private CommandLine() {
		throw new IllegalStateException("Utility class");
	}

	/**
	 * Runs a shell/cmd command and returns a string with the standard output and error
	 */
	public static String runCommand(String command) {
		String output;
		Process proc;
		try {
			proc = Runtime.getRuntime().exec(command);
		} catch (IOException e) {
			throw new VideoControllerException("Can't execute command: " + command);
		}

		InputStreamReader isr = new InputStreamReader(proc.getInputStream());
		BufferedReader br = new BufferedReader(isr);
		InputStreamReader eisr = new InputStreamReader(proc.getErrorStream());
		BufferedReader ebr = new BufferedReader(eisr);
		StringWriter sw = new StringWriter();
		try {
			while ((output = br.readLine()) != null)
				sw.append(output + "\n");
			while ((output = ebr.readLine()) != null)
				sw.append(output + "\n");
		} catch (IOException e) {
			throw new VideoControllerException("Can't get standard output from command: " + command);
		}
		return sw.toString();
	}

	// Temporal methods, to be included later in the portable component

	public static void fileDelete(String fileName, boolean throwIfNotExists) {
		try {
			File f = new File(fileName);
			if (f.exists())
				FileUtils.delete(f);
			else if (throwIfNotExists)
				throw new VideoControllerException("File to delete does not exist: " + fileName);
		} catch (IOException e) {
			throw new VideoControllerException(e);
		}
	}

	public static void fileCopy(String fileFrom, String fileTo) {
		try {
			FileUtils.copyFile(new File(fileFrom), new File(fileTo));
		} catch (IOException e) {
			throw new VideoControllerException("Can't copy " + fileFrom + " to " + fileTo);
		}
	}

	public static boolean fileExists(String fileName) {
		File f = new File(fileName);
		return f.exists();
	}

	public static boolean isAbsolute(String fileName) {
		return fileName.startsWith("/")
				|| fileName.length() >= 3 && fileName.charAt(1) == ':' && fileName.charAt(2) == '/';
	}

}
